File: Code/State_Machine_Testing/python/events.py
from utilities import Utilities
from tokens import Tokens

class Events():
    def __init__(self):
        # Event names mapped with types
        self.num_of_events = Utilities.Random.rand_num(0, len(Tokens().event_names))
        self.events = {}
        self.__generate_events()
    #Eve#---------------------------------------------------------------------------------------------------------------------------------------------#       
    def __generate_events(self):
        # Get event names list
        event_names = Utilities.Lists.get_list(Tokens().event_names)

        # Pick a random event name
        event_name = Utilities.Random.rand_choice(event_names)

        # Picks random Variables from list
        variables = Utilities.Lists.get_list(Tokens().variables)
        variable = Utilities.Random.rand_choice(variables)

        # Generate and save event names
        for _ in range(self.num_of_events):
            # Makes each event have a random type (if it has one)
            variable = Utilities.Random.rand_choice(variables) if Utilities.Random.rand_num(0, 1) else  ""

            # Make sure the chosen event name is not already used
            event_name = Utilities.Variable_Handler.avoid_redecloration(event_name, self.events, event_names)

            #Adds used event name to list
            self.events[event_name] = variable    
    #Eve#---------------------------------------------------------------------------------------------------------------------------------------------#
    # Generates random Event(s)
    def generate_events(self):
        # String to write events in
        code = ""

        # Randomly decides if we have events at all
        if Utilities.Random.rand_num(0, 1):
            # Generates event initialization syntax
            for key, value in self.events.items():
                code += f"event {key} {{{value}}}; \n"
        code += '\n'

        # Return events syntax
        return code
    #Eve#---------------------------------------------------------------------------------------------------------------------------------------------#
    # Get events
    def get_events(self):
        return self.events
    #Eve#---------------------------------------------------------------------------------------------------------------------------------------------#

File: Code/State_Machine_Testing/c++/events.cpp
#include "events.h"

using std::vector, std::string;

Events::Events()
{
    _tokens = _Tokens.getTokens();
    makeNamesAndTypes();
}
void Events::makeNamesAndTypes()
{   
    // List of Variable types and Names
    vector<string> allNames = _tokens[Indices::EVENT_NAMES];

    // // Shuffle the list of names to randomize their order.
    std::shuffle(allNames.begin(), allNames.end(), std::default_random_engine(std::random_device()()));
    
    // How many events to randomly generate based on how many names avaliable
    int numOfEvents = rand_between(1, Indices::EVENT_NAMES);
    
    // The list is shuffled, so no need to check for uniquness
    vector<string> selectedNames(allNames.begin(), allNames.begin() + numOfEvents);

    // Push Variable name and Type pair into the vector
    for (const string& name : selectedNames)
    {
        string type = "";
        if (rand_between(0, 1))
            type = _tokens[Indices::VARIABLES][rand_between(0, _tokens[Indices::VARIABLES].size() - 1)];
        _events.push_back(std::make_pair(name, type));   
    }
}
Events::~Events()
{
}